hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -177,11 +177,6 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
 		return reconcile.Result{}, fmt.Errorf("error validating autoscaling parameters: %w", err)
 	}
 
-	// Default input: if release image is empty set it to the latest hostedCluster
-	if nodePool.Spec.Release.Image == "" {
-		nodePool.Spec.Release.Image = hcluster.Status.Version.History[0].Image
```

Comment:
I find unfortunate to force users to set this explicitly for new nodePools when most of the time they will want the current cluster version.
What if we just add the additional checks to avoid the panic when this is not set. Thoughts?

Reply:
Also if we choose to drop the defaulting we should make this required https://github.com/openshift/hypershift/blob/main/api/v1alpha1/nodepool_types.go#L60

Reply:
The CLI sets the default for you, and I would expect providers (i.e. not using the CLI) to be managing their HostedCluster and NodePools to be automatically generating these anyway (e.g. terraform or some templating system), propagating down values from some other higher level configuration, so I'm not sure who would be burdened.

Re: continuing to check... maybe, but at some point I think it's reasonable to rely on db constraints for basic data validation. Our API now ensures you can't persist an empty value for this field, so I'm not sure about adding guard clauses everywhere the value might be referenced from now on.

Actually I could see an argument that the code should panic if the field is empty now because that's according to our API an invalid state that we shouldn't process because the behavior would be undefined...

Reply:
>Also if we choose to drop the defaulting we should make this required >https://github.com/openshift/hypershift/blob/main/api/v1alpha1/nodepool_types.go#L60

Yeah I think the field is required

Reply:
Marking `.spec.release` required didn't actually change the openAPI schema so it must already be effectively required

Reply:
>The CLI sets the default for you, and I would expect providers (i.e. not using the CLI) to be managing their HostedCluster and NodePools to be automatically generating these anyway (e.g. terraform or some templating system), propagating down values from some other higher level configuration, so I'm not sure who would be burdened.

This is reasonable. Though does the CLI know current cluster version at any point in time? Otherwise any automation would need to have logic to find it from the HC.Status and pass it in to any new created nodePool.

Dropping the defaulting sounds reasonable to me as far as we make it required here https://github.com/openshift/hypershift/blob/main/api/v1alpha1/nodepool_types.go#L60

>Actually I could see an argument that the code should panic if the field is empty now because that's according to our API an invalid state that we shouldn't process because the behavior would be undefined...

This is an interesting perspective, I usually think of API validation/defaulting as purely UX and I expect the backend to handle any error gracefully but never panic.

Reply:
>This is reasonable. Though does the CLI know current cluster version at any point in time? Otherwise any automation would need to have logic to find it from the HC.Status and pass it in to any new created nodePool.

There is no requirement that the NodePool release image be the same as the HostedCluster. It's entirely up to the end user. Maybe they want it to follow the HostedCluster status, or spec, or not at all.

However, this did make me realize the commit breaks the `create nodepool` command, which I had forgotten about and for which we have no CI coverage. Either `--release-image` has to be introduced as a required argument to `create nodepool`, or the command should default to the release image of the associated hostedcluster (or both). So I need to fix that.

/hold

>>Actually I could see an argument that the code should panic if the field is empty now because that's according to our API an invalid state that we shouldn't process because the behavior would be undefined...

>This is an interesting perspective, I usually think of API validation/defaulting as purely UX

I suggest that from our perspective, etcd taken together with the apiserver is a database, and the apiserver enforces the database field level constraints. If we can't rely on that I'd like to understand why, because the implications would be immense (e.g. weaving application level object validation code throughout every point of access duplicating the openAPI validations).

>and I expect the backend to handle any error gracefully but never panic.

I think panicking the entire process is probably too much — I would revise my statement to suggest that if it appears the database level constraints have been violated somehow (again, how?), ignore _that particular resource_ with an error and consider it tainted, but again I'm not sure where the line is for how to achieve that without adding some generic validation pre-reconcile to every controller which reuses the openAPI validators to double-check the persistence layer (effectively dropping trust of the apiserver). i.e., why only check this field to ensure it's valid according to the database constraints and not every single field on all our objects?

Reply:
>There is no requirement that the NodePool release image be the same as the HostedCluster. It's entirely up to the end user. Maybe they want it to follow the HostedCluster status, or spec, or not at all.

true, though I suspect (may be wrongly) that they will want it to follow the HostedCluster status for new nodePools most of the time, thus my biased towards providing a happy path for that from a UX pov.